net: mqtt: BSD socket based implementation

rlubos opened the pull request:
Hello everyone,

This PR introduces new MQTT implementation, based on MQTT from Nordics nRF5 SDK. The library have been reworked, to fit into Zephyr, introducing changes proposed in #8975, which include:

* transport independent MQTT logic, with support for multiple transports (TCP socket and TLS socket transports provided, utilizing recently added TLS socket implementation),
* support for multiple MQTT versions (3.1.0 and 3.1.1 supported),
* single event handler - no need to keep callback array in RAM,
* automatic send of Ping Requests, for connection keep-alive,
* message/event parameters wrapped into strucutres - easier extension for future MQTT versions,
* no separate thread needed to run MQTT - application only needs to call `mqtt_input` and `mqtt_live` periodically

The existing MQTT example (`mqtt_publisher`) was ported to the new API. It was tested successfully on qemu_x86 (with TLS as well) and one of Nordic platforms (not present in this PR). Remaining boards supporting the sample were also verified that they do compile, but they'd require testing though.

MQTT tests were ported as well.

I'm attaching presentation from the October Networking Forum (it can be find on the forum's Google Drive as well):
[mqtt_zephyr.pdf](https://github.com/zephyrproject-rtos/zephyr/files/2442474/mqtt_zephyr.pdf)


Feedback appreciated :)

rlubos committed:
net: mqtt: Add BSD socket implementation

Add new, socket based MQTT implementation, based on MQTT from Nordic
nRF5 SDK, introducing the following features:

* transport independent MQTT logic, with support for multiple transports
* support for multiple MQTT versions (3.1.0 and 3.1.1 supported)
* single event handler - no need to keep callback array in RAM
* automatic send of Ping Requests, for connection keep-alive
* message/event parameters wrapped into strucutres - easier extension
  for future MQTT versions
* no separate thread needed to run MQTT - application only needs to call
  mqtt_input and mqtt_live periodically

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: mqtt: Add TLS socket transport

Add TLS transport to socket MQTT implementation.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
samples: mqtt: Switch MQTT Publisher to new MQTT implementation

Port MQTT Publisher sample to new socket MQTT API.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
tests: mqtt: Port tests to use new implementation

Port existing MQTT tests to new socket MQTT API.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
Some numbers from building `mqtt_publisher`:

**Old MQTT lib:**
qemu_x86:
```
Memory region         Used Size  Region Size  %age Used
	     ROM:       79684 B      4092 KB      1.90%
	     RAM:        112 KB      8188 KB      1.37%
	IDT_LIST:         488 B         2 KB     23.83%
	MMU_LIST:         104 B         1 KB     10.16%
```
frdm_k64f:
```
Memory region         Used Size  Region Size  %age Used
	   FLASH:       62404 B         1 MB      5.95%
	    SRAM:       34288 B       192 KB     17.44%
	IDT_LIST:         136 B         2 KB      6.64%
```	
frdm_k64f_tls:
```
Memory region         Used Size  Region Size  %age Used
	   FLASH:      121328 B         1 MB     11.57%
	    SRAM:       85328 B       192 KB     43.40%
	IDT_LIST:         136 B         2 KB      6.64%
```
**New MQTT lib:**
qemu_x86:
```
Memory region         Used Size  Region Size  %age Used
	     ROM:       84228 B      4092 KB      2.01%
	     RAM:        108 KB      8188 KB      1.32%
	IDT_LIST:         488 B         2 KB     23.83%
	MMU_LIST:         104 B         1 KB     10.16%
```
frdm_k64f:
```
Memory region         Used Size  Region Size  %age Used
	   FLASH:       63208 B         1 MB      6.03%
	    SRAM:       34384 B       192 KB     17.49%
	IDT_LIST:         136 B         2 KB      6.64%
```
frdm_k64f_tls:
```
Memory region         Used Size  Region Size  %age Used
	   FLASH:      122524 B         1 MB     11.68%
	    SRAM:       76720 B       192 KB     39.02%
	IDT_LIST:         136 B         2 KB      6.64%
```

rlubos commented:
CC @carlescufi @laperie

jukkar commented on include/net/mqtt_socket.h:
```diff
@@ -0,0 +1,655 @@
+/*
```
The commit subject "net: mqtt: Add MQTT socket implementation" contains redundant information, what about just saying "net: mqtt: Add BSD socket implementation"

rlubos replied:
Good point, will change.

jukkar commented on include/net/mqtt_socket.h:
```diff
@@ -0,0 +1,655 @@
+} mqtt_suback_return_code_t;
+
+/**@brief MQTT client forward declaration @ref mqtt_client_t for details. */
+typedef struct mqtt_client_t mqtt_client_t;
```
We typically do not use typedef for structs and enums in zephyr.

rlubos replied:
I wasn't sure if it's a requirement to avoid `typedef` in Zephyr, so I've left it as it was originally. Of course I can change that as it's requested.

jukkar commented on include/net/mqtt_socket.h:
```diff
@@ -0,0 +1,655 @@
+	 */
+	mqtt_evt_cb_t evt_cb;
+
+	/** MQTT protocol version. */
```
It looks like there would be holes in the struct, what about ordering the fields so that the u8_t and the bitfields are at the end.

jukkar commented on subsys/net/lib/CMakeLists.txt:
```diff
@@ -3,6 +3,7 @@ add_subdirectory_if_kconfig(lwm2m)
 add_subdirectory_if_kconfig(sntp)
 add_subdirectory_ifdef(CONFIG_DNS_RESOLVER      dns)
 add_subdirectory_ifdef(CONFIG_MQTT_LIB          mqtt)
+add_subdirectory_ifdef(CONFIG_MQTT_SOCKET_LIB   mqtt_socket)
```
Not sure about the MQTT_SOCKET prefix, what about calling it just MQTT and renaming the old mqtt as MQTT_LEGACY? The directory name we can change later when old MQTT library is removed but renaming config variable name is annoying later as people are using the old name.

rlubos replied:
I didn't want to touch the existing implementation, until there was a positive feedback on the new one. Renaming it to MQTT_LEGACY sounds like a good idea though.

GAnthony commented on include/net/mqtt_socket.h:
```diff
@@ -0,0 +1,655 @@
+#ifndef MQTT_H_
+#define MQTT_H_
+
+#include <zephyr/types.h>
```
I recall the original Nordic SDK MQTT source used stdint.h types, which made the code a bit more portable.

Any reason we cannot continue to use stdint.h, as was done here:
e37c3b32a49b6e5ccba1da86ee8ceaa38f97c081 ?

rlubos replied:
That's correct, source implementation used `stdint.h` types. I've changed them as I've thought Zephyr prefers to use `u8_t` etc, especially that checkpatch throws a warning when these standard types are used.
I have nothing against to `stdint.h` types though, so if we agree that it's fine to use it, I'll be happy to bring them back. @jukkar?

rlubos committed:
net: mqtt: Mark existing implementation as legacy

Rename existing headers and sybols to mqtt_legacy, to allow new
implementation to keep old config and header names.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
@jukkar Covered initial comments, I've got rid of `typedefs` and marked old MQTT as legacy. I've rearranged fields in `struct mqtt_client` as requested, though it didn't seem to make difference (at least for x86).

Additionally, I've added commit with MQTT tests ported to the new API.

jukkar commented on subsys/net/lib/mqtt_socket/mqtt_encoder.c:
```diff
@@ -0,0 +1,781 @@
+#define MQTT_3_1_0_PROTO_DESC_LEN 6
+#define MQTT_3_1_1_PROTO_DESC_LEN 4
+
+const u8_t mqtt_3_1_0_proto_desc_str[MQTT_3_1_0_PROTO_DESC_LEN] = {
```
Could this be static as I could not find any access to it outside of this source file?

jukkar commented on subsys/net/lib/mqtt_socket/mqtt_encoder.c:
```diff
@@ -0,0 +1,781 @@
+const u8_t mqtt_3_1_0_proto_desc_str[MQTT_3_1_0_PROTO_DESC_LEN] = {
+	'M', 'Q', 'I', 's', 'd', 'p'
+};
+const u8_t mqtt_3_1_1_proto_desc_str[MQTT_3_1_1_PROTO_DESC_LEN] = {
```
same for this symbol

jukkar commented on subsys/net/lib/mqtt_socket/mqtt_encoder.c:
```diff
@@ -0,0 +1,781 @@
+	'M', 'Q', 'T', 'T'
+};
+
+const struct mqtt_utf8 mqtt_3_1_0_proto_desc = {
```
here too

jukkar commented on subsys/net/lib/mqtt_socket/mqtt_encoder.c:
```diff
@@ -0,0 +1,781 @@
+	.utf_strlen = MQTT_3_1_0_PROTO_DESC_LEN
+};
+
+const struct mqtt_utf8 mqtt_3_1_1_proto_desc = {
```
and here

jukkar commented on subsys/net/lib/mqtt_socket/mqtt_transport_socket_tcp.c:
```diff
@@ -0,0 +1,121 @@
+	MQTT_TRC("Closing socket %d", client->transport.tcp.sock);
+
+	ret = close(client->transport.tcp.sock);
+	if (close < 0) {
```
Do you mean `ret < 0`here?

rlubos replied:
Wow, I'm surprised such a mistake slipped in. Correcting it rightaway.

jukkar reviewed (commented):
There is lot of code and I missed probably lot of things in this review. This needs more eyes but looks very good overall.

rlubos commented:
@jukkar Thank you for the comments, I've addressed them.

pfalcon commented:
> Feedback appreciated :)

So, quick feedback - your today's presentation at the networking forum meeting had such a nice feature list, why can't we have it here in the description?

pfalcon commented on samples/net/mqtt_publisher/src/main.c:
```diff
@@ -294,111 +325,86 @@ static int try_to_connect(struct mqtt_client_ctx *client_ctx)
+				    k_uptime_get();
+		}
+
+		qos++;
```
Ah, so that's why there's a `>` operation against an enum value above. Well, worth a comment there.

And generally, relying on a particular enum value ordering/nearness may be a good optimization trick for internal library code, but looks somewhat hacky in a user-facing sample.

pfalcon commented on samples/net/mqtt_publisher/src/main.c:
```diff
@@ -294,111 +325,86 @@ static int try_to_connect(struct mqtt_client_ctx *client_ctx)
-		k_sleep(APP_SLEEP_MSECS);
-		PRINT_RESULT("mqtt_tx_publish", rc);
-	}
+	while (i++ < 4 * APP_MAX_ITERATIONS && connected) {
```
Ok, my guess after ~2min: because you try 4 different QoS types per iteration?

rlubos replied:
Given you concerns, I think I'll rework the application code a little bit, to be something more similar to the original version. I indeed might be over-optimized for the sample code at this point.

pfalcon reviewed (changes requested):
Some suggestions, looking mostly only at the sample code.

pfalcon commented on include/net/mqtt.h:
```diff
@@ -0,0 +1,629 @@
+
+/**@brief Abstracts UTF-8 encoded strings. */
+struct mqtt_utf8 {
+	u8_t *utf_str;          /**< Pointer to UTF-8 string. */
```
Suggestion: use `utf8`.

pfalcon commented on include/net/mqtt.h:
```diff
@@ -0,0 +1,629 @@
+/**@brief Abstracts UTF-8 encoded strings. */
+struct mqtt_utf8 {
+	u8_t *utf_str;          /**< Pointer to UTF-8 string. */
+	u32_t utf_strlen;       /**< Length of UTF string. */
```
Suggestion: use `len`.

pfalcon commented on include/net/mqtt.h:
```diff
@@ -0,0 +1,629 @@
+
+/**@brief Abstracts binary strings. */
+struct mqtt_binstr {
+	u8_t *bin_str;          /**< Pointer to binary stream. */
```
Suggestion: use `bin` or `data`.

pfalcon commented on include/net/mqtt.h:
```diff
@@ -0,0 +1,629 @@
+/**@brief Abstracts binary strings. */
+struct mqtt_binstr {
+	u8_t *bin_str;          /**< Pointer to binary stream. */
+	u32_t bin_strlen;       /**< Length of binary stream. */
```
Suggestion: use `len`.

pfalcon commented:
What I like in this PR is that in doesn't introduce threads and stuff (again, this feature was presented in @rlubos slides and not very obvious here, hope that will change).

Ping to @GAnthony re: our discussion that protocol implementations should avoid usage of threads and stuff whenever possible (leaving the decision to use or not use threads to applications).

Also ping to @d3zd3z that people actually implement MQTT without using threads ;-). (Didn't review how well it works yet, but I doubt that threads would make it better, both in terms of clarity and resource usage.)

rlubos commented on samples/net/mqtt_publisher/src/main.c:
```diff
@@ -294,111 +325,86 @@ static int try_to_connect(struct mqtt_client_ctx *client_ctx)
+	while (i++ < 4 * APP_MAX_ITERATIONS && connected) {
+		iter_start = k_uptime_get();
+
+		if (qos > MQTT_QoS_2_EACTLY_ONCE) {
```
Lower case "o" is typically found in Quality of Service acronym, I guess that's the reason it's here. It might be upper case though, for consistency.

pfalcon replied:
> It might be upper case though, for consistency.

+1 for consistency. If everything is upper-case, then CamelCase acronyms should be upper-case too.

rlubos commented on samples/net/mqtt_publisher/src/main.c:
```diff
@@ -20,204 +19,172 @@
 
-		mbedtls_ssl_conf_cert_profile(&ctx->tls.mbedtls.conf,
-				&mbedtls_x509_crt_profile_default);
+#if defined(MBEDTLS_KEY_EXCHANGE__SOME__PSK_ENABLED)
```
Well, I'm not 100% happy with these mbedTLS defines here. I don't find myself a Kconfig expert, but I believe it should be doable to have such symbols defined at Kconfig level, depending on selected configuration. We can add it to the wishlist, if mebdTLS Kconfig rework will take place.

rlubos commented on samples/net/mqtt_publisher/src/main.c:
```diff
@@ -246,46 +213,110 @@ static char *get_mqtt_topic(void)
+	param.message.topic.qos = qos;
+	param.message.topic.topic.utf_str = (uint8_t *)get_mqtt_topic();
+	param.message.topic.topic.utf_strlen =
+			strlen(param.message.topic.topic.utf_str);
```
MQTT operates on byte-length, so `utf_strlen` is in bytes as well. I can rework the naming, to avoid confusion.

rlubos commented:
Thanks @pfalcon for the review. I'll rework the code and sample and push the changes soon.

pfalcon commented:
> Thanks @pfalcon for the review. I'll rework the code and sample and push the changes soon.

Thanks! But please again, find time to link your yesterday's presentation, and I'd suggest to add bullet-list from it to the description here. (I also hope the presentation will be available in the usual Network Forum gdrive folder.)

lemrey commented on subsys/net/lib/mqtt_socket/mqtt_encoder.c:
```diff
@@ -0,0 +1,781 @@
+	return err_code;
+}
+
+int connect_request_encode(const struct mqtt_client *client,
```
Perhaps this function would look better if it simply returned on error, instead of executing all code inside one or multiple "if not error" statement. If you need to run some generic error handling code before returning, you can jump to it using a goto.

rlubos replied:
Sure, I agree, but if you look into other parts of the code you'll see, that it's just how the original library was written - there's such a logic all across the code. So to be consistent, it would require changes all accros the library, and it's IMHO something more than a library port, as you start to mess with logic.
Of course, as now we have some nice test coverage in Zephyr also ported, it might be resonable to rework the code in a way you describe, as there is a chance to catch any potential bugs introduced during rework. And the code would get more "Zephyrish" look.

jukkar replied:
IMHO the internals can be changed a bit later if needed, most important thing is to agree on the MQTT APIs so that we do not need to change them a lot later.

rlubos commented:
Pushed an updated version. Review coments are addressed, application reworked slightly to remove obscuring loops and branch is rebased to master. Due to the rebase, I've reworked the code to use the new logger as well.

GAnthony commented:
> transport independent MQTT logic, with support for multiple transports

if I recall the original Nordic code, the transports were used for quite disparate types of transport: sockets versus something like lwip?   
Though it might be a nice feature, I wonder if we really need to use it to distinguish TLS vs non-TLS sockets, given the socket API for both has been unified in Zephyr.   Then if we do it for mqtt, should we do the same for other protocols?  

>Some time later, transport abstraction is there, because guess what - it's useful! 

@pfalcon, Actually, not sure it is really necessary in this case.

pfalcon commented:
> @pfalcon, Actually, not sure it is really necessary in this case.

Necessary is probably not the right word here, nothing beyond assembler (machine code) is necessary. But we rely on the layers to abstract/generalize things, as they're *useful*.

rlubos commented:
@GAnthony 
> if I recall the original Nordic code, the transports were used for quite disparate types of transport: sockets versus something like lwip?

Yes, you're correct, lwip was one of the options.

> Though it might be a nice feature, I wonder if we really need to use it to distinguish TLS vs non-TLS sockets, given the socket API for both has been unified in Zephyr. 

Well, the truth is, the transport are pretty similar, still, they differ in terms of socket creation. They could have a common code-base, with some extra configuration for secure socket. Let me know if you want me to rework it in this PR.

rlubos commented:
I think I'll remove the RFC tag, it's been almost 3 weeks the PR is up, so I guess we can move towards finalizing it. Just ping me if anyone wants to keep it.

GAnthony commented:
> Well, the truth is, the transport are pretty similar, still, they differ in terms of socket creation. They could have a common code-base, with some extra configuration for secure socket. Let me know if you want me to rework it in this PR.

Well, I have no really compelling argument to request changing it.  
Thanks for work.

jukkar commented on subsys/net/lib/mqtt_socket/mqtt.c:
```diff
@@ -0,0 +1,699 @@
+ * @brief MQTT Client API Implementation.
+ */
+
+#define LOG_MODULE_NAME mqtt
```
I suggest that we prefix this with "net_" like in other network related libraries, that way it is possible to set runtime logging filters by just "net*" that would catch all the network related output. Please do that for all the other .c files in this PR too.

jukkar reviewed (changes requested):
Overall looks good, noticed one issue with logging changes that needs to be addressed.

jukkar commented:
@pfalcon are you ok with the changes?

pfalcon commented:
> @pfalcon are you ok with the changes?

I didn't even review the source of the lib (only samples).

If there're reasons to haste with it too much, please share them.

Otherwise, I think it's worth to leave this to be eyeballed my more people for now.

rlubos commented:
@jukkar Thanks, I've aligned the logger prefixes.

@pfalcon The PR was up for over 3 weeks, and actually at this point it would be beneficial for us at Nordic to have this merged. We use it in internal project as a dependency, and having it present in Zephyr would make life a little easier for us. I don't want to push too much, as I understand that the more people will look into it, the better the outcome will be. Just wanted to clarify on our motivations here.

pfalcon commented:
> as I understand that the more people will look into it, the better the outcome will be

I'd say we don't even shoot for that, but just to make sure that as many interested parties as possible to get to know that such a PR exists and give a chance to look inside before it's merged, to avoid getting in something which obviously won't work for someone. In this regard, I'm glad there's a communication with #10500.

Myself, I'll try to finish review by Mon.

pfalcon commented on include/net/mqtt.h:
```diff
@@ -0,0 +1,629 @@
+ *
+ * @retval 0 or an error code indicating reason for failure.
+ *
+ * @note Default protocol revision used for connection request is 3.1.0. Please
```
Copy-paste, please remove.

pfalcon commented on include/net/mqtt.h:
```diff
@@ -0,0 +1,629 @@
+ *       set client.protocol_version = MQTT_VERSION_3_1_1 to use protocol 3.1.1.
+ */
+int mqtt_publish(struct mqtt_client *const client,
+		 struct mqtt_publish_param const *const param);
```
Please use "const struct mqtt_publish_param", here and below.

pfalcon commented on include/net/mqtt.h:
```diff
@@ -0,0 +1,629 @@
+		 struct mqtt_publish_param const *const param);
+
+/**
+ * @brief API used by subscribing client to send acknowledgment to the broker.
```
Can we reword this? Something like "API used by client to send acknowledgment on receiving QoS1 message"...

pfalcon commented on include/net/mqtt.h:
```diff
@@ -0,0 +1,629 @@
+ *
+ * @retval 0 or an error code indicating reason for failure.
+ *
+ * @note Default protocol revision used for connection request is 3.1.0. Please
```
copy-paste

pfalcon commented on include/net/mqtt.h:
```diff
@@ -0,0 +1,629 @@
+ *                   Shall not be NULL.
+ * @param[in] param Identifies message being acknowledged.
+ *
+ * @retval 0 or an error code indicating reason for failure.
```
Btw, it's unclear whether a "reason for failure" is an MQTT status, or global E* statuses. Suggested to clarify (everywhere).

pfalcon commented on include/net/mqtt.h:
```diff
@@ -0,0 +1,629 @@
+ * @note Default protocol revision used for connection request is 3.1.0. Please
+ *       set client.protocol_version = MQTT_VERSION_3_1_1 to use protocol 3.1.1.
+ */
+int mqtt_publish_ack(struct mqtt_client *const client,
```
Also, could go user-friendly and call it e.g. mqtt_publish_qos1_ack(), likewise for qos2.

rlubos replied:
Not sure about it, if we had for instance QoS1 and QoS2 ack messages, but here ack is QoS1-bound, same with others.

pfalcon replied:
> but here ack is QoS1-bound, same with others

I'm not sure I understood the response. If here ack is QoS1-bound, then we can add that fact to the function name, and that's what I propose.

pfalcon commented on include/net/mqtt.h:
```diff
@@ -0,0 +1,629 @@
+
+/**
+ * @brief API used by subscribing clients to acknowledge reception of a
+ *        released message. Should be used on reception @ref MQTT_EVT_PUBREL
```
"released message" or "message release"? Mention again that it's for qos2?

pfalcon commented on include/net/mqtt.h:
```diff
@@ -0,0 +1,629 @@
+			 struct mqtt_pubrec_param const *const param);
+
+/**
+ * @brief API to used by publishing client to request releasing published data.
```
"published data" or "published message"?

pfalcon commented on include/net/mqtt.h:
```diff
@@ -0,0 +1,629 @@
+			  struct mqtt_pubcomp_param const *const param);
+
+/**
+ * @brief API to request subscribing to a topic on the connection.
```
It's "topics", right?

pfalcon replied:
("one or more topics" sounds good)

pfalcon commented on include/net/mqtt.h:
```diff
@@ -0,0 +1,629 @@
+		   struct mqtt_subscription_list const *const param);
+
+/**
+ * @brief API to request unsubscribe from a topic on the connection.
```
Again, "topics".

pfalcon commented on include/net/mqtt.h:
```diff
@@ -0,0 +1,629 @@
+
+/**
+ * @brief API to send MQTT ping.
+ *
```
Perhaps, mention that this is optional, and lib will send it itself based on interval specified in ...

pfalcon replied:
(xref to mqtt_live()).

pfalcon commented on include/net/mqtt.h:
```diff
@@ -0,0 +1,629 @@
+
+/**
+ * @brief API to abort MQTT connection.
+ *
```
Define "abort".

pfalcon replied:
Also, put normal "disconnect" first.

pfalcon commented on include/net/mqtt.h:
```diff
@@ -0,0 +1,629 @@
+ *
+ * @retval 0 or an error code indicating reason for failure.
+ *
+ * @note This API is only supported when using the socket transport layer.
```
Is this not relevant?

pfalcon reviewed (changes requested):
31 comment, just for .h, mostly for docstrings.

pfalcon commented on subsys/net/lib/mqtt_socket/mqtt_encoder.c:
```diff
@@ -0,0 +1,784 @@
+ * @param[in] message_type Message type containing packet type and the flags.
+ *                         Use @ref MQTT_MESSAGES_OPTIONS to construct the
+ *                         message_type.
+ * @param[in] length Buffer where the message payload along with variable
```
Impossible that length is a Buffer.

pfalcon commented on subsys/net/lib/mqtt_socket/mqtt_encoder.c:
```diff
@@ -0,0 +1,784 @@
+ * @param[inout] packet Pointer to the MQTT message variable header and payload.
+ *                      The 5 bytes before the start of the message are assumed
+ *                      by the routine to be available to pack the fixed header.
+ *                      However, since the fixed header length is variable
```
That's creative.

pfalcon reviewed (changes requested):
Review for mqtt_decoder.c/mqtt_encoder.c .

@mike-scott: Did you see this? As I told you, everyone will "read/write flat buffers" in their own way. But I didn't really meant in such way! This examplifies all things which can go wrong with such a code:

1. For parsing/generation, only 2 pointers are required, *cur and *top (or *limit, or *end), where top is initialized to to `buf + len`. You do `cur < top` to check that you aren't out of bounds, and do cur++ as you go. These are conveniently packed into structure like:

~~~
struct buf_ctx {
u8_t *cur;
u8_t *top;
}
~~~

And pointer to that structure is passed to parsing/gen functions. You can compare to what we have here - buf, len, offset, all that marshalled around as flat params.

2. Again, good design precalculates needed length(s) of stuff and encodes varints as it goes, in one pass. Here it's only one varint is used, and close to the beginning of pkt, so the trick is used of preallocating worst-case length for varint, then offsetting start of pkt on 2nd pass.


@rlubos: Sorry for using your PR as a case study of how not to do it ;-). I know it makes no sense to complain about it, because the response is already given "that's how original lib did". There're still comments to fix.

pfalcon commented on subsys/net/lib/mqtt_socket/mqtt.c:
```diff
@@ -0,0 +1,699 @@
+#define TOTAL_BUFFER_COUNT (BUFFER_COUNT_PER_CLIENT * MQTT_MAX_CLIENTS)
+
+/** MQTT Client table. */
+static struct mqtt_client *mqtt_client[MQTT_MAX_CLIENTS];
```
Oh, so all this is for mqtt_live() to not take a param? Well, how about it taking a param and removing O(n) arrays?

pfalcon commented on subsys/net/lib/mqtt_socket/mqtt.c:
```diff
@@ -0,0 +1,699 @@
+	}
+
+	/* Free the instance. */
+	client_free(client);
```
Both looks weird to have it before event_notify(client, ...) and not needed, as you pass  MQTT_EVT_FLAG_INSTANCE_RESET reset there (though is that flag is again needed? why not do it straight and free client after event delivery?).

pfalcon commented on subsys/net/lib/mqtt_socket/mqtt_rx.c:
```diff
@@ -0,0 +1,182 @@
+		err_code = packet_length_decode(data + start, datalen - start,
+						&remaining_length, &offset);
+		if (err_code != 0) {
+			return datalen;
```
Excuse me, what? If we had error during msg len decoding, then we've done with decoding our data?

Well, no. We decoded 0, if the varint length is not complete (packet_length_decode() should return -EAGAIN or something). Or if varint is visibly hosed, *then* it's -EINVAL, and we abort connection (or what does the spec say?).

pfalcon commented on subsys/net/lib/mqtt_socket/mqtt_decoder.c:
```diff
@@ -0,0 +1,342 @@
+		length += (buffer[index] & MQTT_LENGTH_VALUE_MASK) * multiplier;
+		multiplier *= MQTT_LENGTH_MULTIPLIER;
+
+	} while ((buffer[index++] & MQTT_LENGTH_CONTINUATION_BIT) != 0);
```
Can't do it like that. 4 bytes of varint max. If more, kaboom.

pfalcon commented on subsys/net/lib/mqtt_socket/mqtt_rx.c:
```diff
@@ -0,0 +1,182 @@
+		u32_t packet_length = offset + remaining_length;
+
+		if (packet_length > MQTT_MAX_PACKET_LENGTH) {
+			/* We receiving data we cannot handle. */
```
So, abort a connection. Well, you could skip all that data and send an app a notification that too big a packet received, but that's an obvious DoS hole.

pfalcon commented on subsys/net/lib/mqtt_socket/mqtt_rx.c:
```diff
@@ -0,0 +1,182 @@
+
+		if (start + packet_length > datalen) {
+			/* We may want to save this packet for later use.
+			 * This means we received a truncated packet.
```
Well, comment is misleading. It means that a complete MQTT msg wasn't received, so we should keep receiving it (accumulating its data).

pfalcon reviewed (changes requested):
So, this implementation is broken. It's not broken-broken is a traditional way we had it with native API implementations, and few-lines changes might even recover it, albeit in not the most efficient way. The proper solution of course should use explicit state machine.

@rlubos, when doing something with TCP, one should picture 2 cases:

1. TCP packets containing 1 byte of payload each.
2. One big TCP packet packing together a few of app-level protocol messages.

p.2 Seems to be handled, yeah. Bu there're holes in handling of p.1.

jukkar commented on subsys/net/lib/mqtt_socket/mqtt_decoder.c:
```diff
@@ -0,0 +1,342 @@
+		MQTT_TRC(">> BL:%08x, B:%p, O:%08x A:%08x", buffer_len, buffer,
+			 *offset, available_len);
+
+		if (available_len >= SIZE_OF_UINT8) {
```
Good point, I suggest that SIZE_OF_UINT8 is replaced by sizeof(u8_t) and SIZE_OF_UINT16 is replaced by sizeof(u16_t), there should be no need to use a define here especially as the sizeof is evaluated at compile time anyway.

pfalcon commented on subsys/net/lib/mqtt_socket/mqtt_internal.h:
```diff
@@ -0,0 +1,439 @@
+
+/**@brief Verifies if MQTT Client's state is set with one indicated in 'STATE'.
+ */
+#define MQTT_VERIFY_STATE(CLIENT, STATE) ((CLIENT)->state & (STATE))
```
I don't anything here which "verifies" the state. Given existing SET/RESET vocab around, MQTT_IS_STATE or MQTT_HAS_STATE would sound good.

pfalcon commented:
@rlubos, can elaborate on the purpose of mqtt_mutex_lock(), what is it supposed to protect against?

pfalcon commented on subsys/net/lib/mqtt_socket/mqtt_transport_socket_tcp.c:
```diff
@@ -0,0 +1,124 @@
+	int ret;
+
+	while (offset < datalen) {
+		ret = send(client->transport.tcp.sock, data + offset,
```
Oh, so you used "non-blocking recv, blocking send" trick. Cute. Actually, can't even find a fault - the send may take longer than ping interval, but send is *activity*, so there's not need to do anything during send. Of course, individual chunks of send may take longer than ping interval, but then we're really slow and server should and will disconnect us.

But you should document explicitly that publish() and all other client->server operations are blocking, so user can't have more than one MQTT client per thread (should use multiple threads for that).

rlubos replied:
I had an impression that non-blocking  sending on sockets is not fully supported in Zephyr?

pfalcon replied:
Nope, no specific limitations are known. There's https://github.com/zephyrproject-rtos/zephyr/issues/5842, but of course, that's a little bit different than "non-blocking sending on sockets".

But non-blocking send() is not going to help you. Suppose, you were called for publish(), and had a partial send(). What do you do? You either have to return to caller, but then app can ask to publish again, but you didn't even finished with previous! Or you could do your own sub-event loop, to keep recving incoming msg, but then you call app back, and it can ask you to send, again. That's that recursive calling back which I mentioned, which we can't allow, so we'll be inventing ways to preclude app to publish/etc while another send is in progress (why I asked about mutex), or queue them.

All is worse than what there's now. The current impl is smart. Well, it's dumb, and *that's* smart ;-). Exactly what we need for a small RTOS. Just need to deal with some omissions in handling and generalize API.

pfalcon commented on subsys/net/lib/mqtt_socket/Kconfig:
```diff
@@ -0,0 +1,48 @@
+	int "Maximum MQTT packet size"
+	default 128
+	help
+	  Maximum MQTT packet size that can be sent (including the fixed and
```
This doesn't say the whole truth. Not just "sent", but also "received".

rlubos commented on include/net/mqtt.h:
```diff
@@ -0,0 +1,629 @@
+ *
+ * @note The implementation assumes TCP module is enabled.
+ *
+ * @note By default the implementation uses MQTT version 3.1.0.
```
Perhaps it was not well established when library was created, sound reasonable to default to 3.1.1 now.

pfalcon commented:
But now I think I did the majority of the review, and I believe I now understand the logic/algorithm behind it. I'd like to add some comment, whose initial purpose is to structure my own thoughts, but some definitely might call for changes in this PR.

So, first of all, if I would implement an MQTT client, I would think that its purpose would to alleviate a user application of dealing with "MQTT internal" things like QoS1, QoS2. I.e. there would be publish() method, which should behave the same whether there's QoS0, 1, or 2. And that behavior would be "return to user when the message is published", with for higher QoSes opens a tarpit of possible mutual recursion between app and lib, given that MQTT protocol is full-duplex.

But indeed, why? Who said that MQTT is a generic publish-subscribe protocol? No, it's an adhoc protocol with its peculiarities, and if users chose it, they should be ready to deal with those peculiarities. So, I can only applaud shoving all the complexity of handling QoS1/2 from the MQTT client to the application. Overall complexity is indeed reduced, and there's always a choice of not using higher QoSes (e.g. last time I checked, Amazon IoT didn't support QoS2, and from a quick google-up, that still appears to be the case).

rlubos commented on include/net/mqtt.h:
```diff
@@ -0,0 +1,629 @@
+	/** Event result. For example, MQTT_EVT_CONNACK has a result code
+	 *  indicating success or failure code of connection procedure.
+	 */
+	int result;
```
It can contain error code with the reason of the failure, updated the description.

pfalcon replied:
> It can contain error code with the reason of the failure, updated the description.

Well, I mean that if it's MQTT lib's error, it could fit into u8_t/i8_t. If it's E*, int sounds good to not over-optimize.

rlubos commented on include/net/mqtt.h:
```diff
@@ -0,0 +1,629 @@
+	struct mqtt_utf8 *user_name;
+
+	/** Password (if any) to be used for the connection. Note that if
+	 *  password is provided, user name shall also be provided. NULL
```
Will check on that once I'm done with rework.

pfalcon replied:
I checked and appears to be true.

pfalcon replied:
> I recently was told that's not true. Empty password ("") and no password (just username) are 2 different cases, reportedly supported by the standard.

So, any follow-up to this conversation?

rlubos commented on include/net/mqtt.h:
```diff
@@ -0,0 +1,629 @@
+ * @note If more than one simultaneous client connections are needed, please
+ *       define MQTT_MAX_CLIENTS to override default of 1.
+ * @note Please define MQTT_KEEPALIVE time to override default of 1 minute.
+ * @note Please define MQTT_MAX_PACKET_LENGTH time to override default of
```
Well, unless we use dynamic allocation for every packet received, I don't see a way to get around maximum packet length? Especially, that given the MISRA-C recommendations Zephyr is recenly trying to be compliant with, we should avoid it.

pfalcon replied:
I posted a proposal at https://github.com/zephyrproject-rtos/zephyr/pull/10125#issuecomment-429809548 .

I'm not sure that references to MISRA are at the place here.

rlubos commented:
Pushed initial batch of fixes, covering the first part of review.

pfalcon commented:
Now another question, for which I again would like to call to @mike-scott as a witness.

So, last week we discussed with Mike an issue of buffer allocation. The best way to represent it would be a pseudo-dialog (pseudo because no Mike's words appear there, it's just my interpolation of what I heard from him!):

* Mike said that he'd like to know buffer size for recv packet
* Me: But why, you're dealing with COAP, a UDP protocol, just allocate an MTU size, it's not that much
* Mike said he's not satisfied, as it's too bloated an approach


So, with this PR, we deal with TCP, streaming protocol. There's no easy upper bound on packet sizes, unlike UDP. But people aren't too shy to use the same solution as I proposed to Mike (and again, the only reason I proposed it is because size are bounded by pretty small number (== many packets will actually be up to that size)).


Well, no, I don' think we should do it here. It would be easy to say: "we can deal with it later", but I'd invoke @jukkar's words that the main thing for this PR is to settle on the MQTT API, and doing it right would require modifications to the API. So, I think we should do it now.

As sends are blocking already, dealing with publishing data is easy. @rlubos, there's no need to copy message payload into buffer, it can be sent straight from the buffer provided by user.

Dealing with receiving messages is what would require API changes. @rlubos, you would need to pause after you've read headers (both fixed and var) of the received PUBLISH msg, and post the event to the app at this point. The event should include the payload length. Then you would need to provide a func like mqtt_read_payload(), which would have standard stream read() semantics. This function would use the underlying socket in blocking manner (no big deal, as the lib is not async, and if there's a message to receive, the app would have to receive it anyway). The app would call that func one or more time to read the payload (into its own buffers).


That's the nature of my proposal.

rlubos commented:
> @rlubos, can elaborate on the purpose of mqtt_mutex_lock(), what is it supposed to protect against?

Well, while the sample code is a single-threaded application, I can imagine some folks might use it in different way, running the MQTT processing on a separate thread, and sending messages from different one. Hence the mutex.

Regarding QoS and all of the ACK handling by the application. Sure, it could be hidden from the application in some extend. But for instance, MQTT 5.0 already extends all of the ACK messages with the result code or user properties, so it leaves an opening for an application to fill these.

rlubos commented:
@pfalcon Sorry, but I'm not a fan of the changes proposed, due to serveral reasons:

1. I don't understand why do we suddenly drop the asynchronous behavior of the library. While you got me, that `send` was called in a blocking way (but still, I believe that non-blocking implementation of sockets will return an error currently in case it should block, and `poll` implementation is also only a placeholder for now, correct me if I'm wrong), I'd rather fix that than drop the asynchronous behavior of the library.
2. The solution you proposed solves the issue of a big payloads. But still, application needs to allocate/preallocate memory for the payload, and in case it's really big, it will be a major part of the entire message, so we could just provide the buffer to the library from the application (which is pretty much current behavior). The only benefit I can see is that we could read and process payload in chunks (but it's very application specific, the question is if it would make sense for that specific use case).
3. While the payload of the PUBLISH message is application-specific, there are mesasges that have predefined payload format (currently there's one for MQTT client - SUBACK, it gets slightly more complicated in case of server, but that's not the case). Therefore every single application needs to parse it manually, unless we differentiate on behavior between messages.
4. Fragmentation. We focus on large payloads, but the truth is, that low-power IoT devices tend to avoid that. Instead, the important part is to send data in as small number of packets as possible. And when we start to push data in chunks to the lower layer, it will be sent in chunks with current implementation. So for instance to send a simple PUBLISH message, we will need to call send 5 times - fixed header, topic length, topic, payload length, payload (correct me if I misunderstood you proposal here), which will result in 5 packets, each bringing TCP/IP/L2 overhead.

And as for the CoAP - I'm not 100% sure what Mikes intention were so I don't want to speak in his favor, but from my point of view, it's all about the optimization, due to the CoAP nature. In MQTT we have convenience of the TCP protocol - we send and forget. In case of CoAP, it needs to keep the data for possible retransmissions, so if we start to send messages in bulk and be reckless about the memory, we can easily run out of it.

So summarizing, I'd rather to keep the memory-management behavior as it is, while of course fixing other stuff pointed out in the review. But the topic is open for discussion, so if anyone else wants to pariticipate and present their opinion, please go ahead.

rlubos commented on subsys/net/lib/mqtt_socket/mqtt_decoder.c:
```diff
@@ -0,0 +1,342 @@
+
+#define MQTT_LENGTH_VALUE_MASK       0x7F
+#define MQTT_LENGTH_CONTINUATION_BIT 0x80
+#define MQTT_LENGTH_MULTIPLIER       0x80
```
I'm not sure what exactly do you mean here, `#define MQTT_LENGTH_CONTINUATION_BIT (1 << 7)`?

pfalcon replied:
Nope, I meant `#define MQTT_LENGTH_SHIFT       7` and then use `<<` instead of `*`, like in old good days of C (the code of this PR gives "Java" feeling in many places).

Of course, that's relatively minor comparing to other stuff, ~ personal opinion actually.

pfalcon commented:
> Well, while the sample code is a single-threaded application, I can imagine some folks might use it in different way, running the MQTT processing on a separate thread, and sending messages from different one. Hence the mutex.

I see, we discussed such a case with @d3zd3z. My usual response is that different threads should not share I/O resources. And if they want, then *they* (multiple threads) should bother with synchronizing access. Because why burden normal case with one thread == one resource otherwise.

pfalcon commented:
> Regarding QoS and all of the ACK handling by the application. Sure, it could be hidden from the application in some extend.

I assume you mean "to some extent". My experience shows that trying to alleviate app from that leads to unproportional increase of MQTT client complexity (more here: https://github.com/zephyrproject-rtos/zephyr/pull/10125#discussion_r225196335) . So, I think this hits just the right spot.

pfalcon commented:
> I don't understand why do we suddenly drop the asynchronous behavior of the library.

Because there was never such support (as my analysis showed).

 > I'd rather fix that than drop the asynchronous behavior of the library.

Trust me, you don't want that. Changes I propose is a piece of cake comparing to writing fully async impl for a full-duplex, not request-response protocol like MQTT ;-). (Again, a thoughtstream of details is here: https://github.com/zephyrproject-rtos/zephyr/pull/10125#discussion_r225196335).

pfalcon commented:
> The only benefit I can see is that we could read and process payload in chunks (but it's very application specific, the question is if it would make sense for that specific use case).

That's not the "only benefit", that should be a founding requirement for designing such a lib.

Receiving an MQTT msg and shoving data to UART (to your screen during debugging or to another device for processing), or to receiving a firmware update via MQTT and writing it chunk by chunk to flash - all these should be baseline usecases for designing a generic MQTT lib (and we had turns on adhoc MQTT lib(s) in Zephyr already, so hopefully time is to do it right).

pfalcon commented:
> While the payload of the PUBLISH message is application-specific
> ...
> unless we differentiate on behavior between messages.

Of course, my proposal applies only to received PUBLISH.

pfalcon commented:
> 4. Fragmentation. We focus on large payloads,

I thought we focus on implementing client for MQTT protocol? It allows payload of 256MB they say.

> but the truth is, that low-power IoT devices tend to avoid that.

Sure. Do you propose to write separate MQTT lib for each power level of IoT devices? What if we just implement support for protocol, and let each particular device decide how it wants to use it?

> Instead, the important part is to send data in as small number of packets as possible. So for instance to send a simple PUBLISH message, we will need to call send 5 times - fixed header, topic length, topic, payload length, payload (correct me if I misunderstood you proposal here), which will result in 5 packets.

??? On which level do you speak now? You won't find any spec which says that "For SOCK_STREAM, each send() results in one TCP packet delivered". Do you speak about hacked-up/broken Zephyr TCP stack? If so, you're right, but the only worry we should have is to fix it. 

Otherwise:

1. Well, fairly speaking, I was talking about sending outgoing PUBLISH in 2 chunks. But you're absolutely right, it can be done in 5 too ;-).
2. There's no need to worry about 5 packets, we have TCP stack to worry for that. (If someone has an urge to worry, then we have the stream layer too. It's not as good as mine from https://github.com/zephyrproject-rtos/zephyr/pull/5985, which had buffering support as an explicit requirement, but of course, missing parts can be backported).
3. What the worry should be about, is how to steal as little as possible memory from application. In this regard, using 5 sends is better than using 2, you'll literally cut the overhead to mere bytes. You'll be receive long stream of kudos from users around the world in this case, and definitely +100 from me too.
4. A case of *receiving* PUBLISH isn't affected by this concern.

pfalcon commented:
> I'm not 100% sure what Mikes intention were so I don't want to speak in his favor

I bring Mike into these discussions because: a) we discuss closely related matters with him regularly; b) he's the only person I know who cares about memory usage *now* (not in some distant future).

And my point of all this is that unless we'll find a common ground on how to deal with such issues as memory allocation, we'll never have a happy OS with happy users.

My idea of happiness you heard - it's "signing up" a "social contract" that "in this OS, memory belongs to the users, not cannibalized by the OS itself". YMMV

rlubos commented:
> Your code provides very poor and inefficient way of memory management. But the main point is that no way implemented in OS/middleware will be smart enough. Only application may know how to do it best, and should be free to do it as dumb (same 80k) or as smart as it wants.

Then a counter proposal would be to provide a buffer for a client during client initialization from the application, so that the library doesn't have to care about memory at all. We talk about dealing with payload, but what about the topic itself? It can be as large as 64K, if we aim spec limits. 

MQTT 5.0 also extends most of the messages with user properties, so there's possibly more large data to deal with in the header. Probably that's the reason it also introduces Maximum Packet Size property. 

> ??? On which level do you speak now? You won't find any spec which says that "For SOCK_STREAM, each send() results in one TCP packet delivered". Do you speak about hacked-up/broken Zephyr TCP stack? If so, you're right, but the only worry we should have is to fix it.

I'm talking about the current implementation not the specs. And unless there's some kind of Nagle algorightm introduced it will work this way.

pfalcon commented:
> Then a counter proposal would be to provide a buffer for a client during client initialization from the application, so that the library doesn't have to care about memory at all. 

Yes, dull dependency injection is the best solution for the absolutely abstract and reusable library. But knowing 2 extremes: a) fully hidden memory allocation with overcommit; b) full dependency injection from outside - we don't necessarily have to choose one of them, there also compromises inbetween.

> We talk about dealing with payload, but what about the topic itself? It can be as large as 64K, if we aim spec limits.

... So I exactly propose an easy compromise for starters - to allow "arbitrarily sized" (== limited by an app, not by us) payload size. If you like that idea, that we both know that we can resolve it for *send* topic too - by doing that "5 send calls" way. But then we'd need to think how to do it for recv case, and there it's of course a bit "less easy". And that's again why I propose an easy to implement compromise(s) allowing for the greatest impact (savings), I have no intention to drive us/this PR down the rabit hole.

> I'm talking about the current implementation not the specs. And unless there's some kind of Nagle algorightm introduced it will work this way.

And I'm talking about (frustrating to me, as you guessed) topic that we all finally chose a higher level Socket API, but still trying to use it as if we had some adhoc 'struct nbuf''s around. To your concern I already gave answer - any issue we have in the network stack implementation, we'll have to fix. (I may add, we really *will have to*, because I know about these issues for years, and have to say, nobody else seems to care, now that people to actually start to use networking for real, we all will have to care.)

pfalcon commented:
@rlubos, And I wished there were comments on the proposed changes for recv side of PUBLISH, that's where API changes happen, that's what we should discuss and decide.

rlubos commented:
Ok so I see the benefit of your proposal - large meessage (or rather large paylaod) can be notified to the application, and the application can decide whether and how to handle it. There's nothing to argue about it, it's a plus.

But what I don't like is that we will introduce two-stage memory management for the library. We still need to provide some memory to the library, so it is able to parse the fixed/variable header. The only alternative I see is to let the application handle the variable header as well, but it's introducing yet another complexity to the app developer, as he needs to know how the data is structured within the messge. And blocking in the MQTT event handler for unspecified (possibly large as we're talking about receiving here, not waiting for buffers) amount of time does not appeal to me as well. 

If I'd build a MQTT-based ecosystem, I'd not expect the client to be able to handle 20K packet (because still, MQTT packetizes data, even it's over a stream) all of sudden, when typically messages are much smaller. For me it's a bad design. You've gained my curosity with the firmware update, so I've tried to find existing applications for this. Quick google allowed me to find two examples of firmware update over MQTT, both handling firmware in chunks (though I'm sure you could find a counter propsal, I've just gone through a few first google results).

But it's only my opinion. That's why I'd really like to hear others opinion as well on this topic.

pfalcon commented:
> There's nothing to argue about it, it's a plus.

Good that we're on the same line here.

> But what I don't like is that we will introduce two-stage memory management for the library.

I wouldn't worry about that and wouldn't look to find very specific formulations here. As long as we're governed by the desire to preserve application's memory for itself, it can be only good. Yeah, it may take additional effort from the app, so sweet spot may be somewhere in the middle.

> We still need to provide some memory to the library, so it is able to parse the fixed/variable header.

And it will be there, and CONFIG_MQTT_MAX_PACKET_LENGTH, renamed, will specify its size. At least I don't request changing that setup (I would, but I don't). Again, the talk is about **only** keeping payload out of that lib buffer.

> And blocking in the MQTT event handler for unspecified (possibly large as we're talking about receiving here, not waiting for buffers) amount of time does not appeal to me as well.

Well, we can't do anything here - receiving large amounts of data may take long time. Perhaps, you could go smart and see if you should send a ping while an app receives a long message (which might have been spooled to buffers on the broker side, so the broker *app* doesn't know there's an activity ongoing. But please don't do this in this version.

> If I'd build a MQTT-based ecosystem, I'd not expect the client to be able to handle 20K packet (because still, MQTT packetizes data, even it's over a stream) all of sudden, when typically messages are much smaller. 

Well, it's an exercise in the liveliness of imagination. You got your example on firmware upgrade (please don't mix up a usecase and a particular implementations of it which you can google up - there's always more than that to it). Another quick one is a meteostation which usually receives short local sensor updates. But once a day/week it chooses to receive a weather forecast update. As you may imagine, it's already a bit different size. And content providers are such providers - what if they wanted to post ~advertisement~ cyclone map JPG in the update (delivered as XML with hexadecimal encoding, who uses base64, anyway)? Surely, we don't want to limit such an innovation? ;-)

> That's why I'd really like to hear others opinion as well on this topic.

Sure. You asked review to be fast, and I tried to prioritize it from my side. I'm all for more people having gone thru this, as I mentioned.

jukkar commented on subsys/net/lib/mqtt_socket/mqtt_encoder.c:
```diff
@@ -0,0 +1,784 @@
+int ping_request_encode(const struct mqtt_client *client, const u8_t **packet,
+			u32_t *packet_length)
+{
+	(void)client;
```
Please use ARG_UNUSED() macro for all these unused variables.

jukkar commented on subsys/net/lib/mqtt_socket/mqtt_rx.c:
```diff
@@ -0,0 +1,182 @@
+			return datalen;
+		}
+
+		u32_t packet_length = offset + remaining_length;
```
Please place variable declarations to start of the function

jukkar commented:
It might be a good idea if the MQTT library would not directly allocate anything but let the application either supply a buffer to use or allow application to supply an allocator that would return things to the library when it needs buffer space. For example similar way as how subsys/net/buf.c:net_buf_append_bytes() does, of course in this case there would be no net_buf's.

pfalcon reviewed (dismissed):
For (my own too) reference, summary of the changes I consider important to make (and which I'm firm on):

1. publish() should send payload data out of user-supplied buffer, without copying it to lib buffer first.
2. Receiving of incoming PUBLISH message should leave reading payload data to user, by supplying a wrapper function with a standard read() interface.
3. Global array of clients should go, there should not be inter-dependencies among different client instances. 
3.1. Which means that mqtt_live() should take mqtt client pointer like any other func.
3.2. And if there should be mutex there at all, then it belong to the client structure, to global scope.

rlubos commented:
@pfalcon Thanks for the summary. Sorry for sluggish progress from my side, I had some internal interrupts that slowed me down.

I'll push an update soon with reworked encoder/decoder, according to your suggestions (though I've left the fixed header preallocation, which in my opinion is just fine). Then I'll rework the RX path and stuff from point 3.

As for 1 and 2, let's make it a final step.

pfalcon commented:
> Sorry for sluggish progress from my side, I had some internal interrupts that slowed me down.

Sure, np. It's just that I'm back to digging in the other matters too, so thought a summary checklist of previous discussion would help.

> though I've left the fixed header preallocation, which in my opinion is just fine

Ack. My concern is just the payload, to avoid arbitrary limits there right from the start, and design API with that in mind.

rlubos commented:
Pushed further fixes and rewok of encoder/decoder. Further fixes might be delayed due to ELCE next week, I'll add DNM tag.

rlubos commented:
Hey, it took a while, but finally I was able to address almost all remaining comments, including:
* removed client array and maximum client limitation,
* grouped internal fields from client structure into separate struct, for clarity. Moved mutex in there.
* removed memory management code from the library, now application provides buffers needed,
* reworked code responsible for processing of bytes received, to (hopefully) simplify it,
* some minor cleanups that I found reasonable to do during the rework.

What is remaining is to implement API changes related to payload processing, requested by @pfalcon, I'll focus on this now.

pfalcon commented:
@rlubos : Ack, thanks, wanted to ping you about status too.

Well, I'd prefer to re-review when everything will be ready (I'm in the middle of bunch of other things anyway). So, please keep the ticket posted.

rlubos committed:
tests: mqtt: Add test to verify publish reception

Add MQTT test to verify PUBLISH message reception.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
@pfalcon @jukkar I think that at this point I've covered all feedback from the review. I've added an API to read the PUBLISH payload manually, please verify if I understood the idea correctly and implemented it as expected.

Additionally to the API changes, I've added another test app to verify PUBLISH message payload processing, as neither sample app nor existing tests covered it.

Recent changes were put in separate commits to simplify the review process. I think they could be squased once agreed that they are OK.

jukkar reviewed (approved):
Thanks Robert, LGTM

pfalcon commented on subsys/net/lib/mqtt_sock/mqtt.c:
```diff
@@ -627,3 +597,43 @@ int mqtt_input(struct mqtt_client *client)
+	mqtt_mutex_lock(client);
+
+	if (client->internal.remaining_payload == 0) {
+		ret = -EPERM;
```
This is perfectly normal condition, EOF reached. Just return EOF (== length of 0).

pfalcon commented on subsys/net/lib/mqtt_sock/mqtt.c:
```diff
@@ -121,54 +122,17 @@ static int client_connect(struct mqtt_client *client)
-		/* No data received. Try again later. */
-		return 0;
+	if (client->internal.remaining_payload > 0) {
+		return -EPERM;
```
This can't be the right error, can it?

pfalcon replied:
Well, notionally, this is "state [transition] error". Would need to go thru errno's and see what's the closest to it - maybe nothing, POSIX errnos suck :-(

EINPROGRESS is probably more close to sockets, and thus confusing (is that *socket* error that we got or *mqtt lib* error?). EBUSY sounds good then, yeah.

rlubos replied:
Just realized why it came up to me to use `EPERM` here, beacuse it doesn't make much sense basing on Zephyr's `errno.h`. In Zephyr we have:
```
EPERM 1		/* Not owner */
```
While errno man pages state:
```
EPERM           Operation not permitted (POSIX.1-2001).
```
IMO the two descriptions are completely different, and the latter one makes this choice a little more reasonable, as `mqtt_input` shall not be called when payload is available to read. Anyway, if there's a better choice, let's change it.

pfalcon replied:
Well, EPERM is "Operation not permitted", regardless of what Zephyr's bugs in the description say. And "not permitted" in the sense of "permission", i.e. access control, as the lexical form of EPERM clearly says. Regardless of any attempts to clarify its meaning/associate additional constraints, people will keep thinking of it as "there's not enough permission to do that".

For a classical reference (rant) on that matter, see http://blog.unclesniper.org/archives/2-Linux-programmers,-learn-the-difference-between-EACCES-and-EPERM-already!.html - calls like that are futile, things should be named right, to be used right.

(Note that explanation there slightly favors your usage - but not really, from my reading of it. It says EPERM should be used based on the static property of the system, like hardlinks to directory are forever not permitted, whereas we here talk about dynamic property, i.e. "in the current state, that operation is not permitted"). Again, it's sad that POSIX doesn't provide generic "invalid state" error :-(.

pfalcon commented on subsys/net/lib/mqtt_sock/mqtt.c:
```diff
@@ -627,3 +597,43 @@ int mqtt_input(struct mqtt_client *client)
+	if (ret <= 0) {
+		if (ret == 0) {
+			ret = -ENOTCONN;
+
```
Spurious empty line? Should be after brace instead?

pfalcon reviewed (commented):
Reviewed new commit "net: mqtt: Handle publish payload separately". Yes, the propose idea is implemented right, thanks. There're just minor tweaks to achieve POSIX-like semantics required.

Re-reviewing all together is on TODO...

rlubos commented:
@pfalcon Thanks, comments have been adressed.

pfalcon commented on include/net/mqtt.h:
```diff
@@ -1,474 +1,685 @@
+/** @brief Abstracts UTF-8 encoded strings. */
+struct mqtt_utf8 {
+	u8_t *utf8;             /**< Pointer to UTF-8 string. */
+	u32_t size;             /**< Size of UTF string, in bytes. */
```
I still find it unfortunate that "mqtt_utf8" has "size", while "mqtt_binstr" - "len".

(Not requesting to fix anything.)

pfalcon replied:
For example, all of struct mqtt_puback_param, struct mqtt_pubrec_param, struct mqtt_pubrel_param, struct mqtt_pubcomp_param managed to end up with the member named message_id, but for some reason, mqtt_utf8 and mqtt_binstr have got to have it different.

pfalcon commented on include/net/mqtt.h:
```diff
@@ -1,474 +1,685 @@
+	struct mqtt_connack_param connack;
 
-int mqtt_connect(struct mqtt_ctx *ctx);
+	/** Parameters accompanying MQTT_EVT_PUBLISH event. */
```
Should we say here that payload field is to be ignored? Should we define separate struct for received PUBLISH?

rlubos replied:
I'll extend the PUBLISH event description.

pfalcon commented on include/net/mqtt.h:
```diff
@@ -1,474 +1,685 @@
+ * @note Default protocol revision used for connection request is 3.1.1. Please
+ *       set client.protocol_version = MQTT_VERSION_3_1_0 to use protocol 3.1.0.
+ * @note If more than one simultaneous client connections are needed, please
+ *       modify :option:`CONFIG_MQTT_MAX_CLIENTS` to override default of 1.
```
A consequence of the good design - one less option for users to care about! ;-) (Please remove.)

pfalcon commented on include/net/mqtt.h:
```diff
@@ -1,474 +1,685 @@
+ *       modify :option:`CONFIG_MQTT_MAX_CLIENTS` to override default of 1.
+ * @note Please modify :option:`CONFIG_MQTT_KEEPALIVE` time to override default
+ *       of 1 minute.
+ * @note Please modify :option:`CONFIG_MQTT_MAX_PACKET_LENGTH` time to override
```
Likewise.

pfalcon commented on include/net/mqtt.h:
```diff
@@ -1,474 +1,685 @@
- * @retval -EINVAL
- * @retval -ENOMEM
- * @retval -EIO
+ * @note Shall be called before connecting the client in order to avoid
```
Please mention that this should be called even before setting other fields in the client. I may suggest: "Shall be called to initialize client structure, before setting any client parameters and before connecting to broker."

pfalcon commented on subsys/net/lib/mqtt_sock/mqtt.c:
```diff
@@ -0,0 +1,639 @@
+
+	if (client->internal.remaining_payload == 0) {
+		ret = 0;
+		goto error;
```
Please-please let's change this to "goto out", or "done", or "finish", just not "error".

pfalcon reviewed (changes requested):
Great work, @rlubos! I think that we have very solid and efficient MQTT library here, thanks for going thru all the refactoring. We will likely need to clarify/cleanup some behavioral aspects yet later. What's really left here is updating docstrings to not contain info no longer relevant after refactoring, clarify init behavior, and - I hope - to rename a got label. I'll -1 for now to avoid spurious merge. I'll be ready to +1 immediately after that.

My own checklist is below:

* [x]    publish() should send payload data out of user-supplied buffer, without copying it to lib buffer first.
* [x]    Receiving of incoming PUBLISH message should leave reading payload data to user, by supplying a wrapper function with a standard read() interface.
* [x]    Global array of clients should go, there should not be inter-dependencies among different client instances.
* [x]    3.1. Which means that mqtt_live() should take mqtt client pointer like any other func.
* [x]    3.2. And if there should be mutex there at all, then it belongs to the client structure, not global scope.

rlubos commented:
@pfalcon Thank you for the feedback, I've addressed final remarks.

As the API changes have beed reviewed and I can see that there is a chance to have this PR merged soon, I've squashed the commits with the API change.

rlubos commented:
And rebased to current master.

pfalcon reviewed (approved):
Thanks!
